src/event/reader_group.rs

```diff
@@ -0,0 +1,507 @@
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// http://www.apache.org/licenses/LICENSE-2.0
+//
+
+use crate::client_factory::ClientFactory;
+use crate::event::reader::EventReader;
+use crate::event::reader_group_state::Offset;
+
+use pravega_client_shared::{Reader, Scope, ScopedSegment, ScopedStream};
+
+use serde::{Deserialize, Serialize};
+use serde_cbor::Error as CborError;
+use serde_cbor::{from_slice, to_vec};
+use snafu::ResultExt;
+use snafu::Snafu;
+use std::collections::HashMap;
+use std::sync::Arc;
+use tokio::sync::Mutex;
+
+cfg_if::cfg_if! {
+    if #[cfg(test)] {
+        use crate::event::reader_group_state::MockReaderGroupState as ReaderGroupState;
+    } else {
+        use crate::event::reader_group_state::ReaderGroupState;
+    }
+}
+
+/// A collection of readers that collectively read all the events in the stream.
+///
+/// The events are distributed among the readers in the group such that each event goes to only one reader.
+///
+/// The readers in the group may change over time. Readers are added to the group by invoking the
+/// [`ReaderGroup::create_reader`] API.
+///
+/// [`ReaderGroup::create_reader`]: ReaderGroup::create_reader
+/// # Examples
+/// ```no_run
+/// use pravega_client_config::ClientConfigBuilder;
+/// use pravega_client::client_factory::ClientFactory;
+/// use pravega_client_shared::{ScopedStream, Scope, Stream};
+///
+/// #[tokio::main]
+/// async fn main() {
+///    let config = ClientConfigBuilder::default()
+///         .controller_uri("localhost:8000")
+///         .build()
+///         .expect("creating config");
+///     let client_factory = ClientFactory::new(config);
+///     let scope = Scope::from("scope".to_string());
+///     let stream = ScopedStream {
+///         scope: scope.clone(),
+///         stream: Stream::from("stream".to_string()),
+///     };
+///     // Create a reader group to read data from the Pravega stream.
+///     let rg = client_factory.create_reader_group(scope, "rg".to_string(), stream).await;
+///     // Create a reader under the reader group.
+///     let mut reader1 = rg.create_reader("r1".to_string()).await;
+///     let mut reader2 = rg.create_reader("r2".to_string()).await;
+///     // EventReader APIs can be used to read events.
+/// }
+/// ```
+pub struct ReaderGroup {
+    pub name: String,
+    config: ReaderGroupConfig,
+    pub state: Arc<Mutex<ReaderGroupState>>,
+    client_factory: ClientFactory,
+}
+
+impl ReaderGroup {
+    // This ensures the mock reader group state object is used for unit tests.
+    cfg_if::cfg_if! {
+        if #[cfg(test)] {
+            async fn create_rg_state(
+            _scope: Scope,
+            _name: String,
+            _rg_config: ReaderGroupConfig,
+            _client_factory: &ClientFactory,
+            _init_segments: HashMap<ScopedSegment, Offset>,
+        ) -> ReaderGroupState {
+            ReaderGroupState::default()
+        }
```

Comment:
Odd indentation. Is there anything we can do about that with `fmt` reverting it?

Reply:
Fixed! Looks like fmt is not applied in this section of code. Probably because of the `cfg_if`?